src/Uno.UWP/UI/Composition/Uno/CompositorThread.android.cs

```diff
@@ -0,0 +1,186 @@
+		//private List<(RenderNode node, double x, double y)> _animatedNodes = new List<(RenderNode node, double x, double y)>();
+
+		private readonly HardwareRenderer _hwRender = new HardwareRenderer();
+		private readonly RenderNode _visualTree = new RenderNode("visual-tree");
+		private readonly ManualResetEventSlim _frameRendered = new ManualResetEventSlim(false);
+		private readonly ICompositionRoot _activity;
+
+		private Thread? _thread;
+
+		public CompositorThread(ICompositionRoot activity)
+		{
+			_activity = activity;
+
+			_hwRender.SetContentRoot(_visualTree);
+			_activity.Window!.TakeSurface(this);
+		}
+		
+		void ISurfaceHolderCallback.SurfaceCreated(ISurfaceHolder? holder)
+		{
+			_hwRender.SetSurface(holder!.Surface);
+
+			Start();
+		}
+
+		void ISurfaceHolderCallback.SurfaceChanged(ISurfaceHolder holder, Format format, int width, int height)
+		{
+			_visualTree.SetPosition(new Rect(new Point(), new Size(width, height)));
+		}
+
+		void ISurfaceHolderCallback.SurfaceDestroyed(ISurfaceHolder? holder)
+		{
+			Stop();
+
+			_hwRender?.Stop();
+			_hwRender?.Destroy(); // Note: This only release the surface, we can restore it by setting a new surface
+		}
+
+		void ISurfaceHolderCallback2.SurfaceRedrawNeeded(ISurfaceHolder? holder)
+		{
+			_frameRendered.Reset();
+			_frameRendered.Wait();
+		}
+
+		void ISurfaceHolderCallback2.SurfaceRedrawNeededAsync(ISurfaceHolder holder, Java.Lang.IRunnable drawingFinished)
+		{
+			_frameRendered.Reset();
+
+			Task.Run(() => // TODO: Should we invoke drawingFinished on calling / UI thread?
+			{
+				try
+				{
+					_frameRendered.Wait();
+					drawingFinished.Run();
+				}
+				catch (Exception e)
+				{
+					this.Log().Error("Async redraw failed.", e);
+				}
+			});
+		}
+
+		private void Start()
+		{
+			if (_thread is {})
+			{
+				this.Log().Warn("Try to start an already running compositor thread.");
+				return;
+			}
+
+			_thread = new Thread(() =>
+			{
+				_hwRender.Start();
+
+				Looper.Prepare(); // Required to get a Choreographer.Instance
+				Choreographer.Instance!.PostFrameCallback(this);
+
+				Looper.Loop(); // Start the looper so the Choreographer.Instance will invoke our callback
+			})
+			{
+				Name = "Compositor",
+				Priority = ThreadPriority.AboveNormal
+			};
+
+			_thread.Start();
+		}
+
+		private void Stop()
+		{
+			try
+			{
+				_thread?.Abort();
+			}
+			catch (Exception e)
+			{
+				this.Log().Warn("Failed to abort compositor thread.", e);
+			}
+			_thread = null;
+		}
+
+		void Choreographer.IFrameCallback.DoFrame(long frameTimeNanos)
+		{
+			RenderFrame(frameTimeNanos);
+
+			Choreographer.Instance!.PostFrameCallback(this);
+		}
+
+		private void RenderFrame(long frameTimeNanos)
+		{
+			RecordingCanvas canvas;
+			try
+			{
+				canvas = _visualTree.BeginRecording();
+			}
+			catch (Java.Lang.IllegalStateException)
+			{
+				// Cannot start recording (most probably recording is already running somehow).
```

Comment:
Is this indicating an unrecoverable problem?

Reply:
Not necessarily, it might be due to a concurrency issue (the `_visualTree` node might already be in recording state from another thread). But I would also agree that it might also indicate an unrecoverable problem ...

In any case this would be invalid, I will add some error logging here.